From ef966ff67eea92fa9a1e34b797dda4ae63b88083 Mon Sep 17 00:00:00 2001
From: James Morse <james.morse@arm.com>
Date: Mon, 22 Jan 2024 14:58:10 +0800
Subject: [PATCH 658/960] downstream: arm_mpam: resctrl: Add
 resctrl_arch_get_config()

Implement resctrl_arch_get_config() by testing the configuration for a
CPOR bitmap. For any other configuration type return the default.

Signed-off-by: James Morse <james.morse@arm.com>
Signed-off-by: Zeng Heng <zengheng4@huawei.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 drivers/platform/mpam/mpam_devices.c |  3 ++
 drivers/platform/mpam/mpam_resctrl.c | 44 ++++++++++++++++++++++++++++
 2 files changed, 47 insertions(+)

diff --git a/drivers/platform/mpam/mpam_devices.c b/drivers/platform/mpam/mpam_devices.c
index 7eb6c3a57aad..79ab28d63e7f 100644
--- a/drivers/platform/mpam/mpam_devices.c
+++ b/drivers/platform/mpam/mpam_devices.c
@@ -2302,6 +2302,9 @@ int mpam_apply_config(struct mpam_component *comp, u16 partid,
 
 	lockdep_assert_cpus_held();
 
+	if (!memcmp(&comp->cfg[partid], cfg, sizeof(*cfg)))
+		return 0;
+
 	comp->cfg[partid] = *cfg;
 	arg.comp = comp;
 	arg.partid = partid;
diff --git a/drivers/platform/mpam/mpam_resctrl.c b/drivers/platform/mpam/mpam_resctrl.c
index 82325956a938..cda8f08c0819 100644
--- a/drivers/platform/mpam/mpam_resctrl.c
+++ b/drivers/platform/mpam/mpam_resctrl.c
@@ -247,6 +247,50 @@ int mpam_resctrl_setup(void)
 	return err;
 }
 
+u32 resctrl_arch_get_config(struct rdt_resource *r, struct rdt_domain *d,
+			    u32 closid, enum resctrl_conf_type type)
+{
+	u32 partid;
+	struct mpam_config *cfg;
+	struct mpam_props *cprops;
+	struct mpam_resctrl_res *res;
+	struct mpam_resctrl_dom *dom;
+	enum mpam_device_features configured_by;
+
+	lockdep_assert_cpus_held();
+
+	if (!mpam_is_enabled())
+		return r->default_ctrl;
+
+	res = container_of(r, struct mpam_resctrl_res, resctrl_res);
+	dom = container_of(d, struct mpam_resctrl_dom, resctrl_dom);
+	cprops = &res->class->props;
+
+	partid = resctrl_get_config_index(closid, type);
+	cfg = &dom->comp->cfg[partid];
+
+	switch (r->rid) {
+	case RDT_RESOURCE_L2:
+	case RDT_RESOURCE_L3:
+		configured_by = mpam_feat_cpor_part;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	if (!r->alloc_capable || partid >= resctrl_arch_get_num_closid(r) ||
+	    !mpam_has_feature(configured_by, cfg))
+		return r->default_ctrl;
+
+	switch (configured_by) {
+	case mpam_feat_cpor_part:
+		/* TODO: Scaling is not yet supported */
+		return cfg->cpbm;
+	default:
+		return -EINVAL;
+	}
+}
+
 void resctrl_arch_reset_resources(void)
 {
 	int i, idx;
-- 
2.43.0

